### PR TITLE
driver: PS/2: npcx: add driver support for Nuvoton npcx family

### PR DESCRIPTION
**Pull request by ChiHuaL**
The PS/2 module in npcx provides a hardware accelerator mechanism
including an 8-bit shift register, a state machine, and control logic
that handle both the incoming and outgoing data. The hardware
accelerator mechanism is shared by 4 PS/2 channels. To support it,
this CL separates the PS/2 driver into channel and controller drivers.
The controller driver is in charge of the PS/2 transaction. The channel
driver is in charge of the connection between the Zehpyr PS/2 API
interface and controller driver.

Signed-off-by: Jun Lin <CHLin56@nuvoton.com>

### CODE REVIEW

File: drivers/ps2/Kconfig.npcx [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# NPCX PS2 configuration options
+
+# Copyright (c) 2021 Nuvoton Technology Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig PS2_NPCX
+	bool "Nuvoton NPCX embedded controller (EC) PS2 driver"
+	depends on SOC_FAMILY_NPCX && ESPI_PERIPHERAL_8042_KBC
+	help
+	  Enable the NPCX family PS2 driver. The driver also
```
Comment (sjg20) [inline drivers/ps2/Kconfig.npcx:10]: PS/2

Should describe what it supports (keyboard?) and what it does. Needs more detail

File: drivers/ps2/ps2_npcx_channel.c [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+LOG_MODULE_REGISTER(ps2_npcx_channel, CONFIG_PS2_LOG_LEVEL);
+
+#include "ps2_npcx_controller.h"
+
+/* Device config */
+struct ps2_npcx_ch_config {
+	/* pinmux configuration */
+	const uint8_t alts_size;
+	const struct npcx_alt *alts_list;
+	uint8_t channel_id;
```
Comment (sjg20) [inline drivers/ps2/ps2_npcx_channel.c:33]: comment for channel_id

File: drivers/ps2/ps2_npcx_controller.c [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+/* Driver data */
+struct ps2_npcx_ctrl_data {
+	/*
+	 * Bit mask to record the enabled PS/2 channels.
+	 * Only bit[7] and bit[5:3] are used
+	 * (i.e. the bit position of CLK3-0 in the PS2_PSOSIG register)
+	 */
+	uint8_t channel_enabled_mask;
+	struct k_sem lock;
+	struct k_sem tx_sync_sem;
```
Comment (sjg20) [inline drivers/ps2/ps2_npcx_controller.c:53]: missing some member coments

File: drivers/ps2/ps2_npcx_controller.c [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+/*
+ * Set WDAT3-0 and clear CLK3-0 in the PSOSIG register to
+ * reset the shift mechanism.
+ */
+#define NPCX_PS2_SHIFT_MECH_RESET (uint8_t) ~NPCX_PSOSIG_CLK_MASK_ALL
+
+/* in 50us units */
+#define PS2_RETRY_COUNT 10000
+
+#define PS2_TRANSACTION_TIMEOUT K_MSEC(20)
```
Comment (sjg20) [inline drivers/ps2/ps2_npcx_controller.c:36]: 20ms is a long time...can you add a comment as to how it is chosen?

File: drivers/ps2/ps2_npcx_controller.c [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ps2_npcx_ctrl, CONFIG_PS2_LOG_LEVEL);
+
+#define NPCX_PS2_CH_COUNT 4
+
+/*
+ * Set WDAT3-0 and clear CLK3-0 in the PSOSIG register to
+ * reset the shift mechanism.
+ */
+#define NPCX_PS2_SHIFT_MECH_RESET (uint8_t) ~NPCX_PSOSIG_CLK_MASK_ALL
```
Comment (sjg20) [inline drivers/ps2/ps2_npcx_controller.c:31]: Do you need the cast? If so, can you drop the space after it?

**Comment by albertofloyd** [comment]
Update codeowners file

File: drivers/ps2/ps2_npcx_controller.c [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+	struct ps2_reg *const inst = HAL_PS2_INSTANCE(dev);
+
+	/*
+	 * The driver pulls the CLK for non-active channels to low when Start
+	 * bit is detected and pull the CLK of the active channel low after
+	 * Stop bit detected. The EOT bit is set when Stop bit is detected,
+	 * but both SOT and EOT are cleared when all CLKs are pull low
+	 * (due to Shift Mechanism is reset)
+	 */
+	return (IS_BIT_SET(inst->PSTAT, NPCX_PSTAT_SOT) |
```
Comment (sjg20) [inline drivers/ps2/ps2_npcx_controller.c:131]: Should that be || ?

Alternatively, can you use 

IS_BIT_SET(inst->PSTAT, NPCX_PSTAT_SOT | NPCX_PSTAT_EOT) ?

File: drivers/ps2/ps2_npcx_controller.c [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+}
+
+static void ps2_npcx_ctrl_isr(const struct device *dev)
+{
+	uint8_t active_ch, mask;
+	struct ps2_reg *const inst = HAL_PS2_INSTANCE(dev);
+	struct ps2_npcx_ctrl_data *const data = DRV_DATA(dev);
+
+	/*
+	 * ACH = 1 : CHannel 0
```
Comment (sjg20) [inline drivers/ps2/ps2_npcx_controller.c:229]: Channel

File: drivers/ps2/ps2_npcx_controller.c [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+	 * ACH = 2 : CHannel 1
+	 * ACH = 4 : CHannel 2
+	 * ACH = 5 : CHannel 3
+	 */
+	active_ch = GET_FIELD(inst->PSTAT, NPCX_PSTAT_ACH);
+	active_ch = active_ch > 2 ? (active_ch - 2) : (active_ch - 1);
+	LOG_DBG("ACH: %d\n", active_ch);
```
Comment (sjg20) [inline drivers/ps2/ps2_npcx_controller.c:236]: Do you need to range-check active_ch?

File: drivers/ps2/ps2_npcx_controller.c [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+	 */
+	irq_disable(DT_INST_IRQN(0));
+
+	if (channel_id >= NPCX_PS2_CH_COUNT) {
+		LOG_ERR("unexpected channel ID: %d", channel_id);
+		return -EINVAL;
+	}
+
+	if (enable) {
+		data->channel_enabled_mask |= BIT(NPCX_PSOSIG_CLK(channel_id));
```
Comment (sjg20) [inline drivers/ps2/ps2_npcx_controller.c:105]: You use BIT(NPCX_PSOSIG_CLK(channel_id)) a lot

How about adding a function that returns the mask for a particular channel's clock?

File: drivers/ps2/ps2_npcx_controller.c [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+			inst->PSCON &= ~BIT(NPCX_PSCON_XMT);
+			k_sem_give(&data->tx_sync_sem);
+		} else {
+			if (ps2_npcx_ctrl_is_rx_error(dev) == 0) {
+				ps2_callback_t callback;
+				uint8_t data_in = inst->PSDAT;
+
+				LOG_DBG("Recv:0x%02x", data_in);
+				callback = data->callback_isr[active_ch];
+				if (callback != NULL)
```
Comment (sjg20) [inline drivers/ps2/ps2_npcx_controller.c:277]: Is this allowed to be NULL ?

**Review by sjg20 — CHANGES_REQUESTED** [review]
LGTM, some nits

File: drivers/ps2/Kconfig.npcx (hunk 1 above)
Comment (ChiHuaL) [reply drivers/ps2/Kconfig.npcx]: Done.

File: drivers/ps2/ps2_npcx_channel.c (hunk 2 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_channel.c:34]: Done.

File: drivers/ps2/ps2_npcx_controller.c (hunk 3 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c:60]: Done.

File: drivers/ps2/ps2_npcx_controller.c (hunk 4 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c]: Indeed, 20ms is long. Normally, a PS/2 transaction should take about 1.1 ms. Reduce the timeout from 20 ms to 2 ms (1.1 ms + some buffer.)

File: drivers/ps2/ps2_npcx_controller.c (hunk 5 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c]: There is a compile warning if no cast. Drop the space.

File: drivers/ps2/ps2_npcx_controller.c (hunk 6 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c]: Done by changing from "|" to "||".

File: drivers/ps2/ps2_npcx_controller.c (hunk 7 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c]: Done.

File: drivers/ps2/ps2_npcx_controller.c (hunk 8 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c:254]: The active channel number is reported from the hardware module.
It should not report incorrect channel No. unless there is a bug in the hardware.
So I don't check the range of it.

File: drivers/ps2/ps2_npcx_controller.c (hunk 9 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c]: Done by adding a function to return the bit-mask.

File: drivers/ps2/ps2_npcx_controller.c (hunk 10 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c:295]: The callback is assigned after the PS/2 channel is configured.
If someone enables the channel without configuring it first, the callback is NULL.

**Review by sjg20 — APPROVED** [review]

File: drivers/ps2/CMakeLists.txt [hunk 11]
```diff
@@ -3,3 +3,5 @@
 zephyr_library()
 
 zephyr_library_sources_ifdef(CONFIG_PS2_XEC		ps2_mchp_xec.c)
+zephyr_library_sources_ifdef(CONFIG_PS2_NPCX		ps2_npcx_channel.c)
```
Comment (albertofloyd) [inline drivers/ps2/CMakeLists.txt:6]: This PR is missing a change into CODEOWNERS file
/drivers/ps2/*npcx*                        @ChiHual

File: drivers/ps2/ps2_npcx_channel.c [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+#define NPCX_PS2_CHANNEL_INIT_FUNC(inst) _CONCAT(ps2_npcx_channel_init_, inst)
+#define NPCX_PS2_CHANNEL_INIT_FUNC_DECL(inst)                                  \
+	static int ps2_npcx_channel_init_##inst(const struct device *dev)
+
+#define NPCX_PS2_CHANNEL_INIT_FUNC_IMPL(inst)                                  \
+	static int ps2_npcx_channel_init_##inst(const struct device *dev)      \
+	{                                                                      \
+		struct ps2_npcx_ch_data *const data = DRV_DATA(dev);           \
+									       \
+		data->ps2_ctrl = device_get_binding(DT_INST_BUS_LABEL(inst));  \
```
Comment (WealianLiao) [inline drivers/ps2/ps2_npcx_channel.c:119]: #36557 provides a method(DEVICE_DT_GET) to get the controller at build time. We can get the parent node by DT_PARENT.

File: drivers/ps2/CMakeLists.txt (hunk 11 above)
Comment (ChiHuaL) [reply drivers/ps2/CMakeLists.txt:6]: Thanks for the comment.  
I have added Nuvoton developers to the CODEOWNERS file in the 2nd commit of this PR.

File: drivers/ps2/ps2_npcx_channel.c (hunk 12 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_channel.c]: I followed PR #36557 to use DEVICE_DT_GET to get the controller device object. Please see commit #3.

File: drivers/ps2/ps2_npcx_controller.c [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+DEVICE_DT_INST_DEFINE(0, &ps2_npcx_ctrl_init, NULL, &ps2_npcx_ctrl_data_0,
+		      &ps2_npcx_ctrl_config_0, POST_KERNEL,
+		      CONFIG_PS2_INIT_PRIORITY, NULL);
+
+static int ps2_npcx_ctrl_init(const struct device *dev)
+{
+	const struct ps2_npcx_ctrl_config *const config = DRV_CONFIG(dev);
+	struct ps2_npcx_ctrl_data *const data = DRV_DATA(dev);
+	struct ps2_reg *const inst = HAL_PS2_INSTANCE(dev);
+	const struct device *clk_dev = device_get_binding(NPCX_CLK_CTRL_NAME);
```
Comment (WealianLiao) [inline drivers/ps2/ps2_npcx_controller.c:331]: #36557 just merged. Here should modify after rebase. `DEVICE_DT_GET(NPCX_CLK_CTRL_NODE);`

Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c]: Done.

**Review by albertofloyd — APPROVED** [review]

File: drivers/ps2/ps2_npcx_controller.c [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+	k_sem_take(&data->lock, K_FOREVER);
+	/*
+	 * Disable the interrupt during changing the enabled channel mask to
+	 * prevent from preemption.
+	 */
+	irq_disable(DT_INST_IRQN(0));
+
+	if (channel_id >= NPCX_PS2_CH_COUNT) {
+		LOG_ERR("unexpected channel ID: %d", channel_id);
+		return -EINVAL;
```
Comment (keith-zephyr) [inline drivers/ps2/ps2_npcx_controller.c:117]: This path leaves the interrupt disabled and still owns the semaphore.

File: drivers/ps2/ps2_npcx_controller.c [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+int ps2_npcx_ctrl_configure(const struct device *dev, uint8_t channel_id,
+			    ps2_callback_t callback_isr)
+{
+	struct ps2_npcx_ctrl_data *const data = DRV_DATA(dev);
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	if (channel_id >= NPCX_PS2_CH_COUNT) {
+		LOG_ERR("unexpected channel ID: %d", channel_id);
+		return -EINVAL;
```
Comment (keith-zephyr) [inline drivers/ps2/ps2_npcx_controller.c:88]: Need to give back the semaphore or do this check before k_sem_take().

File: drivers/ps2/ps2_npcx_controller.c [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+	}
+
+	if (callback_isr == NULL) {
+		return -EINVAL;
```
Comment (keith-zephyr) [inline drivers/ps2/ps2_npcx_controller.c:92]: Same comment here - move this outside the critical section.

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: drivers/ps2/ps2_npcx_channel.c [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+#define NPCX_PS2_CHANNEL_INIT(inst)                                            \
+									       \
+	static const struct npcx_alt ps2_channel_alts##inst[] =                \
+		NPCX_DT_ALT_ITEMS_LIST(inst);                                  \
+									       \
+	static const struct ps2_npcx_ch_config ps2_npcx_ch_cfg_##inst = {      \
+		.channel_id = DT_INST_PROP(inst, channel),                     \
+		.alts_size = ARRAY_SIZE(ps2_channel_alts##inst),               \
+		.alts_list = ps2_channel_alts##inst,                           \
+		.ps2_ctrl = DEVICE_DT_GET(DT_PARENT(DT_DRV_INST(inst))),       \
```
Comment (mbolivar-nordic) [inline drivers/ps2/ps2_npcx_channel.c:109]: I'm not seeing a call to device_is_ready to check the status of the ps2_ctrl device before using it. Where is it?

File: drivers/ps2/ps2_npcx_channel.c [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+	};                                                                     \
+									       \
+	DEVICE_DT_INST_DEFINE(inst, ps2_npcx_channel_init, NULL, NULL,         \
+			      &ps2_npcx_ch_cfg_##inst, POST_KERNEL,            \
+			      CONFIG_PS2_CHANNEL_INIT_PRIORITY,                \
```
[the remark this answers is not in the document]
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_channel.c:119]: Done.

File: drivers/ps2/ps2_npcx_controller.c (hunk 14 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c:117]: Done by giving the semaphore back before return.

File: drivers/ps2/ps2_npcx_controller.c (hunk 15 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c:86]: Done.

File: drivers/ps2/ps2_npcx_controller.c (hunk 16 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c:90]: Done.

File: drivers/ps2/ps2_npcx_channel.c (hunk 17 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_channel.c:114]: Thanks for pointing out this. 
I added the checking in the function ps2_npcx_channel_init.

File: drivers/ps2/ps2_npcx_controller.c (hunk 14 above)
Comment (keith-zephyr) [reply drivers/ps2/ps2_npcx_controller.c:117]: What about the interrupt?  It is disabled on if channel_id is invalid, but never enabled.

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: drivers/ps2/ps2_npcx_controller.c (hunk 14 above)
Comment (ChiHuaL) [reply drivers/ps2/ps2_npcx_controller.c:118]: Yes, it should be re-enabled before return. I have fixed it. Thanks for catching it.

**Review by keith-zephyr — APPROVED** [review]

**Pull request merged**